Neos.Neos/Classes/Domain/Workspace/WorkspacePublisher.php

```diff
@@ -0,0 +1,283 @@
+        foreach ($changes as $change) {
+            if (
+                !$this->isChangePublishableWithinAncestorScope(
+                    $contentRepository,
+                    $contentStreamId,
+                    $change,
+                    $ancestorNodeTypeName,
+                    $ancestorId
+                )
+            ) {
+                continue;
+            }
+
+            $nodeIdsToPublishOrDiscard[] = new NodeIdToPublishOrDiscard(
+                $change->nodeAggregateId,
+                $change->originDimensionSpacePoint->toDimensionSpacePoint()
+            );
+        }
+
+        return NodeIdsToPublishOrDiscard::create(...$nodeIdsToPublishOrDiscard);
+    }
+
+    private function isChangePublishableWithinAncestorScope(
+        ContentRepository $contentRepository,
+        ContentStreamId $contentStreamId,
+        Change $change,
+        NodeTypeName $ancestorNodeTypeName,
+        NodeAggregateId $ancestorId
+    ): bool {
+        // see method comment for `isChangeWithSelfReferencingRemovalAttachmentPoint`
+        // to get explanation for this condition
+        if ($this->isChangeWithSelfReferencingRemovalAttachmentPoint($change)) {
+            if ($ancestorNodeTypeName->equals(NodeTypeNameFactory::forSite())) {
+                return true;
+            }
+        }
+
+        $subgraph = $contentRepository->getContentGraph()->getSubgraph(
+            $contentStreamId,
+            $change->originDimensionSpacePoint->toDimensionSpacePoint(),
+            VisibilityConstraints::withoutRestrictions()
+        );
+
+        // A Change is publishable if the respective node (or the respective
+        // removal attachment point) has a closest ancestor that matches our
+        // current ancestor scope (Document/Site)
+        $actualAncestorNode = $subgraph->findClosestNode(
+            $change->removalAttachmentPoint ?? $change->nodeAggregateId,
+            FindClosestNodeFilter::create(nodeTypes: $ancestorNodeTypeName->value)
+        );
+
+        return $actualAncestorNode?->nodeAggregateId->equals($ancestorId) ?? false;
+    }
+
+    /**
+     * Before the introduction of the WorkspacePublisher, the UI only ever
+     * referenced the closest document node as a removal attachment point.
+     *
+     * Removed document nodes therefore were referencing themselves.
+     *
+     * In order to enable publish/discard of removed documents, the removal
+     * attachment point of a document MUST refer to an ancestor. The UI now
+     * references the site node in those cases.
+     *
+     * Workspaces that were created before this change was introduced may
+     * contain removed documents, for which the site node can longer be
+     * located, because we have no reference to their respective site.
+     *
+     * Every document node that matches that description will be published
+     * or discarded by WorkspacePublisher::publishSite, regardless of what
+     * the current site is.
+     */
```

Comment:
Thanks for the clarification it took me a bit to understand ^^

But in comparison with https://github.com/neos/neos-ui/pull/3752/commits/768dd8d66896438729868c9ad4faf421e712512c it becomes clear.
It was just a bit tricky as the $ancestorNodeTypeName param is not passed until here but checked in a nested if.

I think we can safely annotated this as `@deprecated` and mark it to be removed before the Neos 9 release?

It would be interesting to know if the publish count will also be calculated correctly despite this and this one change is deletable.

That self referencing _was_ possible in the `removalAttachmentPoint` is imo a flaw of the constraint checks and should be disallowed in the future.

Reply:
I don't think we need to deprecate private methods in a final class

Reply:
yes not to prevent anyone from using it but before releasing well search for `@deprecated` things and cleanup stuff.
I guess we might not want to have / need this hack forever and thus can clean it up after a month or two (and instruct people to check their workspaces)

Reply:
Fine by me, we should at latest be able to remove this once the whole ChangeFinder has been refactored/rewritten

Reply:
done